src/jit/codegenxarch.cpp

```diff
@@ -176,23 +176,63 @@ void                CodeGen::genEmitGSCookieCheck(bool pushReg)
 
     // Make sure that EAX is reported as live GC-ref so that any GC that kicks in while
     // executing GS cookie check will not collect the object pointed to by EAX.
-    if (!pushReg && (compiler->info.compRetType == TYP_REF))
-        gcInfo.gcRegGCrefSetCur |= RBM_INTRET;    
+    if (!pushReg)
+    {
+
+#ifdef FEATURE_UNIX_AMD64_STRUCT_PASSING
+
+        if (varTypeIsStruct(compiler->info.compRetType))
+        {
+            if (compiler->compMethodReturnsMultiRegRetType())
+            {
```

Comment:
On Further thinking, it doesn't seem right to me that codegen figuring out the return type of the method when it is already done by importer and fixed it on GT_RETURN nodes.  The right model to me is that genEmitGSCookieCheck() should have a GenTree* param for which it is generating GS cookie check.  Note that GenTree* node could be GT_RETURN or tail call or Jmp "Method".

The call stack is 
genExitCode(BasicBlock *block) --> genEmitGSCookieCheck()

genExitCode() can pass the last top-level stmtnt of the block as a param to genEmitGSCookieCheck().

When GenTree node is GT_RETURN, there is a need to know the return type.  GT_RETURN will have return type as TYP_STRUCT only when it is a struct returned in multi-regs otherwise it will be a normalized type or a scalar type.